File: trunk/download/stream/test/src/simple/xml/util/ResolverTest.java
package simple.xml.util;

import java.io.StringWriter;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.SortedSet;
import java.util.Set;

import simple.xml.ValidationTestCase;
import simple.xml.load.Persister;
import simple.xml.Attribute;
import simple.xml.Element;
import simple.xml.ElementList;
import simple.xml.Root;

public class ResolverTest extends ValidationTestCase {
        
   private static final String LIST = 
   "<?xml version=\"1.0\"?>\n"+
   "<test name='example'>\n"+
   "   <list>\n"+  
   "      <match pattern='*.html' value='text/html'/>\n"+
   "      <match pattern='*.jpg' value='image/jpeg'/>\n"+
   "      <match pattern='/images/*' value='image/jpeg'/>\n"+
   "      <match pattern='*.exe' value='application/octetstream'/>\n"+
   "      <match pattern='*.txt' value='text/plain'/>\n"+
   "      <match pattern='/html/*' value='text/html'/>\n"+
   "   </list>\n"+
   "</test>";  
   
   @Root(name="match")
   private static class ContentType extends Match {

      @Attribute(name="value")
      private String value;        

      public ContentType() {
         super();                  
      }

      public ContentType(String pattern, String value) {
         this.pattern = pattern;
         this.value = value;        
      }
   }
   
   @Root(name="test")
   private static class ContentResolver implements Iterable<ContentType> {

      @ElementList(name="list", type=ContentType.class)
      private Resolver<ContentType> list;           

      @Attribute(name="name")
      private String name;

      public Iterator<ContentType> iterator() {
         return list.iterator();
      }

      public void add(ContentType type) {
         list.add(type);              
      }

      public ContentType resolve(String name) {
         return list.resolve(name);              
      }

      public int size() {
         return list.size();              
      }
   }
        
	private Persister serializer;

	public void setUp() {
	   serializer = new Persister();
	}
	
   public void testResolver() throws Exception {    
      ContentResolver resolver = (ContentResolver) serializer.read(ContentResolver.class, LIST);

      assertEquals(6, resolver.size());
      assertEquals("image/jpeg", resolver.resolve("image.jpg").value);
      assertEquals("text/plain", resolver.resolve("README.txt").value);
      assertEquals("text/html", resolver.resolve("/index.html").value);
      assertEquals("text/html", resolver.resolve("/html/image.jpg").value);
      assertEquals("text/plain", resolver.resolve("/images/README.txt").value);
      assertEquals("image/jpeg", resolver.resolve("/images/image.JPEG").value);
      
      validate(resolver, serializer);
   }
   
   public void testCache() throws Exception {    
      ContentResolver resolver = (ContentResolver) serializer.read(ContentResolver.class, LIST);

      assertEquals(6, resolver.size());
      assertEquals("image/jpeg", resolver.resolve("image.jpg").value);
      assertEquals("text/plain", resolver.resolve("README.txt").value);
      
      Iterator<ContentType> it = resolver.iterator();
      
      while(it.hasNext()) {
         ContentType type = it.next();
         
         if(type.value.equals("text/plain")) {
            it.remove();
            break;
         }
      }
      resolver.add(new ContentType("*", "application/octetstream"));

      assertEquals("application/octetstream", resolver.resolve("README.txt").value);
      assertEquals("application/octetstream", resolver.resolve("README.txt").value);            
      
      resolver.add(new ContentType("README.*", "text/html"));
      resolver.add(new ContentType("README.txt", "text/plain"));

      assertEquals("text/plain", resolver.resolve("README.txt").value);
      assertEquals("text/html", resolver.resolve("README.jsp").value);

      validate(resolver, serializer);
   }

   public void testNoResolution() throws Exception {
      ContentResolver resolver = (ContentResolver) serializer.read(ContentResolver.class, LIST);

      assertEquals(6, resolver.size());
      assertEquals("text/plain", resolver.resolve("README.txt").value);
      assertEquals(null, resolver.resolve("README"));           
   }

   public void testNonGreedyMatch() throws Exception {
      ContentResolver resolver = (ContentResolver) serializer.read(ContentResolver.class, LIST);

      assertEquals(6, resolver.size());
      resolver.add(new ContentType("/*?/html/*", "text/html"));
      
      assertEquals(7, resolver.size());
      assertEquals(null, resolver.resolve("/a/b/html/index.jsp"));
      assertEquals("text/html", resolver.resolve("/a/html/index.jsp").value);
   }
}

File: port/src/main/Xml/Util/ResolverTest.cs
#region Using directives
using SimpleFramework.Xml.Core;
using SimpleFramework.Xml;
using System.Collections.Generic;
using System;
#endregion
namespace SimpleFramework.Xml.Util {
   public class ResolverTest : ValidationTestCase {
      private const String LIST =
      "<?xml version=\"1.0\"?>\n"+
      "<test name='example'>\n"+
      "   <list>\n"+
      "      <match pattern='*.html' value='text/html'/>\n"+
      "      <match pattern='*.jpg' value='image/jpeg'/>\n"+
      "      <match pattern='/images/*' value='image/jpeg'/>\n"+
      "      <match pattern='/log/**' value='text/plain'/>\n"+
      "      <match pattern='*.exe' value='application/octetstream'/>\n"+
      "      <match pattern='**.txt' value='text/plain'/>\n"+
      "      <match pattern='/html/*' value='text/html'/>\n"+
      "   </list>\n"+
      "</test>";
      [Root(Name="match")]
      private static class ContentType : Match {
         [Attribute(Name="value")]
         private String value;
         [Attribute]
         private String pattern;
         public ContentType() {
            super();
         }
         public ContentType(String pattern, String value) {
            this.pattern = pattern;
            this.value = value;
         }
         public String Pattern {
            get {
               return pattern;
            }
         }
         //public String GetPattern() {
         //   return pattern;
         //}
            return String.format("%s=%s", pattern, value);
         }
      }
      [Root(Name="test")]
      private static class ContentResolver : Iterable<ContentType> {
         [ElementList(Name="list", Type=ContentType.class)]
         private Resolver<ContentType> list;
         [Attribute(Name="name")]
         private String name;
         private ContentResolver() {
            this.list = new Resolver<ContentType>();
         }
         public Iterator<ContentType> Iterator() {
            return list.Iterator();
         }
         public void Add(ContentType type) {
            list.Add(type);
         }
         public ContentType Resolve(String name) {
            return list.Resolve(name);
         }
         public int Size() {
            return list.Size();
         }
      }
   	private Persister serializer;
   	public void SetUp() {
   	   serializer = new Persister();
   	}
      public void TestResolver() {
         ContentResolver resolver = (ContentResolver) serializer.Read(ContentResolver.class, LIST);
         AssertEquals(7, resolver.Size());
         AssertEquals("image/jpeg", resolver.Resolve("image.jpg").value);
         AssertEquals("text/plain", resolver.Resolve("README.txt").value);
         AssertEquals("text/html", resolver.Resolve("/index.html").value);
         AssertEquals("text/html", resolver.Resolve("/html/image.jpg").value);
         AssertEquals("text/plain", resolver.Resolve("/images/README.txt").value);
         AssertEquals("text/plain", resolver.Resolve("/log/access.log").value);
         Validate(resolver, serializer);
      }
      public void TestCache() {
         ContentResolver resolver = (ContentResolver) serializer.Read(ContentResolver.class, LIST);
         AssertEquals(7, resolver.Size());
         AssertEquals("image/jpeg", resolver.Resolve("image.jpg").value);
         AssertEquals("text/plain", resolver.Resolve("README.txt").value);
         Iterator<ContentType> it = resolver.Iterator();
         while(it.hasNext()) {
            ContentType type = it.next();
            if(type.value.equals("text/plain")) {
               it.remove();
               break;
            }
         }
         resolver.Add(new ContentType("*", "application/octetstream"));
         AssertEquals("application/octetstream", resolver.Resolve("README.txt").value);
         AssertEquals("application/octetstream", resolver.Resolve("README.txt").value);
         resolver.Add(new ContentType("README.*", "text/html"));
         resolver.Add(new ContentType("README.txt", "text/plain"));
         AssertEquals("text/plain", resolver.Resolve("README.txt").value);
         AssertEquals("text/html", resolver.Resolve("README.jsp").value);
         Validate(resolver, serializer);
      }
      public void TestNoResolution() {
         ContentResolver resolver = (ContentResolver) serializer.Read(ContentResolver.class, LIST);
         AssertEquals(7, resolver.Size());
         AssertEquals("text/plain", resolver.Resolve("README.txt").value);
         AssertEquals(null, resolver.Resolve("README"));
      }
      public void TestNonGreedyMatch() {
         ContentResolver resolver = (ContentResolver) serializer.Read(ContentResolver.class, LIST);
         AssertEquals(7, resolver.Size());
         resolver.Add(new ContentType("/*?/html/*", "text/html"));
         AssertEquals(8, resolver.Size());
         AssertEquals(null, resolver.Resolve("/a/b/html/index.jsp"));
         AssertEquals("text/html", resolver.Resolve("/a/html/index.jsp").value);
      }
      public void TestResolverCache() {
         ContentResolver resolver = new ContentResolver();
         for(int i = 0; i <= 2000; i++) {
            resolver.Add(new ContentType(String.valueOf(i), String.valueOf(i)));
         }
         AssertEquals(resolver.Resolve("1").value, "1");
         AssertEquals(resolver.Resolve("2000").value, "2000");
      }
   }
}
